File: solutions/c/secret-handshake/1/secret_handshake.c
// Dedicated to Shree DR.MDD — eternal guidance in code

#include "secret_handshake.h"
#include "secret_handshake.h"

#include <assert.h>
#include <stdlib.h>

#define HANDSHAKE_CAPACITY 4

const char **commands(size_t code) {
    const char **steps = calloc(HANDSHAKE_CAPACITY, sizeof(char *));
    size_t idx = 0;

    if ((code & 16) == 16) {
        if ((code & 8) == 8) steps[idx++] = "jump";
        if ((code & 4) == 4) steps[idx++] = "close your eyes";
        if ((code & 2) == 2) steps[idx++] = "double blink";
        if ((code & 1) == 1) steps[idx++] = "wink";
    } else {
        if ((code & 1) == 1) steps[idx++] = "wink";
        if ((code & 2) == 2) steps[idx++] = "double blink";
        if ((code & 4) == 4) steps[idx++] = "close your eyes";
        if ((code & 8) == 8) steps[idx++] = "jump";
    }

    assert(idx <= HANDSHAKE_CAPACITY);
    return steps;
}

File: solutions/r/secret-handshake/1/secret-handshake.R
handshake <- function(n) {
  # Define the available actions and their corresponding bitmask values
  actions <- c("wink", "double blink", "close your eyes", "jump")
  masks <- c(1, 2, 4, 8)
  
  # Identify which actions are present by checking if the bits are set
  selected_actions <- actions[bitwAnd(n, masks) > 0]
  
  # Check if the 5th bit (16) is set to reverse the order of actions
  if (bitwAnd(n, 16) > 0) {
    selected_actions <- rev(selected_actions)
  }
  
  # Return NULL (which is equivalent to c()) if no actions are matched
  if (length(selected_actions) == 0) {
    return(NULL)
  }
  
  return(selected_actions)
}